### **Jeder Programmierer** 

stößt irgendwann auf Schwierigkeiten und fühlt sich vielleicht festgefahren – das ist völlig normal. Programmieren, wie so vieles im Leben, ist anfangs herausfordernd, aber mit der Zeit wird es einfacher.

Um euch das Leben zu erleichtern und eure Leidenschaft fürs Programmieren auch nach der Blockwoche zu bewahren, habe ich einen Leitfaden erstellt. Dieser soll euch in unsicheren Momenten als Orientierungshilfe dienen, damit ihr nachschlagen könnt, wie bestimmte Dinge funktionieren.

Denkt daran: **„Jeder Meister war einmal ein Anfänger.“** Genauso braucht ihr als Anfänger im Programmieren Zeit, um wirklich gut zu werden.

Aber keine Sorge – **ihr schafft das!**


---

# Leitfaden für Python

---

## 1. Python Basic Syntax
- ### 1.1. Variabeln
- ### 1.2. Kommentare
- ### 1.3. Datentypen
- ### 1.4. Operatoren
- ### 1.5. Logische Operatoren
- ### 1.6. Schleifen
    - ### 1.6.1 for loop
    - ### 1.6.2 while
- ### 1.7 if, elif, else   
    - ### 1.7.1 if-Bedigungen
    - ### 1.7.2 else-Bedigungen
    - ### 1.7.3 Elif-Bedigungen
- ### 1.8 Funktionen
## 2. Libraries
- ### 2.1. Understanding Libraries
    - #### 2.1.1 How to read Docs

- ### 2.2. Pandas
- ### 2.2. scikit-learn
- ### 2.3. Plotly
- ### 2.3. Streamlit 


## 3. Fehlerbehebungen

---

So einfach kannst du dein erstes Programm schreiben


In [ ]:
print("Hello World")

`print("Hello World")`

Dabei kann der print befehl dir helfen überhaupt festzustellen ob alles funkoniert.

Solltest du keine Fehler Meldung in dein Terminal oder untern den Block sehen kann es weiter gehen.


## 1.1 Variabeln:
----

- Variabeln können verstanden werden als ein Kontainer für alle möglichen Werte, Daten, Objekte, etc.
- Was du nun in der Variable einspeichert liegt an deinem Code


In [ ]:
my_firstname = "Max"  

my_lastname = "Mustermann"    

print("Deine Varibale enthält den Namen: " + my_firstname)
print("Deine Varibale enthält den Nachnamen: " + my_lastname)

Nun wissen wir das die Variable **`my_firstname`** den Wert enthält den du einträgst und die Variable **`my_lastname`** genauso.
Wir können nun auf diese Variablen zugreifen und diese auch verändern so oft wir wollen.
Dabei sind beide Variablen sogenante *String*.
String referenziert auf ein *Datentypen* dazu mehr in 1.3

## 1.2 Kommentare 
---

**Kommentare** sind in der Programmier Welt sehr wichtig. Sie helfen dabei den Code zu verstehen und zu dokumentieren.
Kommentare werden in Python mit einem # eingeleitet. Alles was nach dem `#` kommt wird nicht ausgeführt.

In [ ]:
# Folgender Code gibt das Alter der Person wieder
Alter = 26 
print("Das Alter beträgt: ", Alter)

## 1.3 Daten Typen
---

#### **Vorhandene Datentypen in Python:**

<span style="color:lightblue"> **int** </span>: 
Eine ganze Zahl, wie `5`, `-3` oder `42`.

<span style="color:lightblue"> **float** </span>: 
Eine Dezimalzahl, wie `3.14`, `-0.001` oder `2.0`.

<span style="color:lightblue"> **str** </span>: 
Ein Text, der in Anführungszeichen steht, wie `"Hallo, Welt!"` oder `'Python ist toll'`.

<span style="color:lightblue"> **bool** </span>: 
Ein Wert, der entweder `True` (wahr) oder `False` (falsch) ist.

In [4]:
# - int (Absoulte Zahlen)
x = 20

# - float (Dezimalzahlen)
y = 20.5

# - str (Strings)
Name = "Das ist ein String der aus Buchstaben besteht und eine Zeichenkette bildet"

# - bool (Booleans)
Male = False
Female = True

<span style="color:lightblue"> **list** </span>: 

Eine Sammlung von Elementen in eckigen Klammern, getrennt durch Kommas, z.B. `[1, 2, 3]` oder `["Apfel", "Banane", "Kirsche"]`.

In [ ]:
Früchte = ["Apfel", "Birne", "Banane", "Kirsche", "Erdbeere"]

print(Früchte)    # Apfel

<span style="color:lightblue"> **Dictionaries** </span>: 

Ein Wörterbuch mit Schlüssel-Wert-Paaren, das in geschweiften Klammern steht, wie 

`{"name": "Alice", "alter": 25}`.

In [ ]:
Lagerbestand = {"Früchte":"Erdbeere",
                "Menge": 40}

Lagerbestand

Dies war ein kleine Einleitung welche Datentypen es gibt.
In den folgeden Kapitel werden wir lernen wie die Datentypen gefiltert, manipuliert werden und vieles mehr

| Mehr über **list** und **dictonaries** findest du in ***Kapitel 1.6*** und ***1.7***|
|-------------------------------------------------------------------------------------|



### 1.4 Operatoren
---

Variablen können auch **kombiniert** werden 

In [ ]:
# Eine kleine Übersicht über verschiedene mathematische Operatoren.
x = 20
y = 10

# Addition
addition = x + y
print("Addition:", addition)

# Subtraktion
subtraktion = x - y
print("Subtraktion:", subtraktion)

# Multiplikation
multiplikation = x * y
print("Multiplikation:", multiplikation)

# Division
division = x / y
print("Division:", division)

# Potenzen
potenz = x ** y
print("Potenz:", potenz)

### 1.5 Logische Operatoren
---

In [ ]:
# and (und)
und_operator = x < 5 and x < 10
print("and Operator:", und_operator)

# or (oder)
oder_operator = x < 5 or x < 4
print("or Operator:", oder_operator)

# not (nicht)
nicht_operator = not(x < 5 and x < 10)
print("not Operator:", nicht_operator)

---
### 1.6 Listen (list)
---
In diesen Abschnitt geht es um die Handhabung mit Listen.

<span style="color:lightblue"> list </span>:

Eine Liste ist eine Datenstruktur in Python, die eine veränderbare, geordnete Sequenz von Elementen darstellt. Jede Eintragung in einer Liste wird als Element bezeichnet. Listen werden durch eckige Klammern [ ] initialisiert.

Eine Liste kann Elemente unterschiedlichen Datentyps enthalten. So kann eine Liste beispielsweise nur Strings oder Zahlen enthalten, aber auch verschiedene Datentypen in einer einzigen Liste kombinieren.

Eine Liste mit <span style="color:lightblue"> **Strings** </span>:

`Früchte = ["Apfel", "Birne", "Banane", "Kirsche", "Erdbeere"]`

Eine Liste mit <span style="color:lightblue"> **Integers** </span>:

`menge = [3, 4, 6, 10, 20]`

Eine Liste mit <span style="color:lightblue"> **Float** </span>:

`Preise = [4.99, 1.99, 6.99, 10.00, 20.95]`

Eine Liste mit <span style="color:lightblue"> **Booleans** </span>:

`verfügbar = [True, False, True, True, False]`

In [ ]:
Früchte = ["Apfel", "Birne", "Banane", "Kirsche", "Erdbeere"]
menge = [3, 4, 6, 10, 20]
Preise = [4.99, 1.99, 6.99, 10.00, 20.95]
verfügbar = [True, False, True, True, False]

print("Früchte: ", Früchte)
print("menge: ", menge)
print("Preise: ", Preise)
print("verfügbar: ", verfügbar)


| Index   | 0      | 1      | 2      | 3       | 4         |
|---------|--------|--------|--------|---------|-----------|
| Element | Apfel  | Birne  | Banane | Kirsche | Erdbeere  |

In [ ]:
Früchte = ["Apfel", "Birne", "Banane", "Kirsche", "Erdbeere"]
print(Früchte[0])   # Apfel Der Index beginnt bei 0 und nicht bei 1
print(Früchte[1])   # Birne an der Index Nummer 1 
print(Früchte[-1])  # Erdbeere an der Index Nummer -1 also das letzte Element in der Liste

#### Mehrere Werte aus einer Liste extrahieren:
<span style="color:lightblue"> **your_list[Inklusiv:Exklusiv]** </span>

Mit diesem Code werden die Elemente 2 bis 4 aus der Liste gefiltert. Dabei ist das 2. Element inklusiv und das 5. Element exklusiv:



In [ ]:
print(Früchte[2:5])  # ['Banane', 'Kirsche', 'Erdbeere'] Der Index 5 wird nicht mitgezählt

Mit diesem Code werden **alle Elemente** bis zur **Index-Nummer 4** aus der Liste gefiltert. 
Dabei wird das 2. Element inklusiv und das 5. Element exklusiv betrachtet:

In [ ]:
print(Früchte[:4])   # ['Apfel', 'Birne', 'Banane', 'Kirsche']  Die ersten 4 Elemente werden ausgegeben

print(Früchte[2:])   # ['Banane', 'Kirsche', 'Erdbeere']        Alle Elemente ab dem Index 2 werden ausgegeben

*WICHTIG:*
- In unseren Fall bedeutet das Banane, Kirsche und Erdbeere
- Das heißt die 2 (Banane) ist inklusiv und die 5 (Erdbeere) ist exklusiv


#### **Werte hinzufügen**

#### **Element zur einer Liste hinzufügen**

In [ ]:
# Element am Ende der Liste hinzufügen
Früchte.append('Ananas')
print("Ananas hinzufügen:" ,Früchte)

#### **Element an einem bestimmten Index hinzufügen**

In [ ]:
# Element an Index Nr. 5 hinzufügen

Früchte.insert(5, 'Orange')
print("Orange an den Index Punkt 5 hinzufügen:" ,Früchte)

#### **Werte ändern:**

In Index Nr 1. Ist der Wert *"Birne"*

`Früchte = ["Apfel", "Birne", "Banane", "Kirsche", "Erdbeere"]`

*"Birne"* wird nun erstzt im darunter liegen den Code durch *'Kiwi'* im Index Nr.1 

In [ ]:
# Element in der Liste ändern

Früchte[1] = 'Kiwi'
print(Früchte)

#### **Werte entfernen:**
- durch "remove" wird der string "Erdbeere" entfernt

In [ ]:
# Element aus der Liste entfernen

Früchte.remove('Erdbeere')
print(Früchte)

In [ ]:
x = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
x.sort()
print(x)

---

### 1.7 Dictionaries 

---


**Ein Dictionary** ist eine Sammlung, um Daten zu speichern, die in Form von Schlüssel-Wert-Paaren vorliegen.
Ein Dictionary üblicherweise in einer Variablen gespeichert. In diesem Beispiel nennen wir die Variable `my_dictionary`.
Ein **Key (Schlüssel)** kann als Spaltenname verstanden werden, und der **Item (Wert)** referenziert den Inhalt, der in dieser "Spalte" stehen soll.

Der allgemeine Aufbau eines Dictionaries sieht wie folgt aus:

##### 1. Möglichkeit
- beachte das vor den Doppelpunkte (:) `Key`eingeleitet wird und nach den (:) das `Iteam` folgt.

In [ ]:
my_dictionary = {
    "Key1": "Item1",
    "Key2": "Item2",
    "Key3": "Item3"
}

my_dictionary

##### 2. Möglichkeit

Die Funktion *`dict`* formatiert automatisch dein Dictionarie im richtig Format

In [ ]:
my_dictionary = dict(
            Key1 = "Item1",
            Key2 = "Item2",
            Key3 = "Item3"
)
my_dictionary

- Key: Dies ist der Name des Schlüssels (z.B. "Key1"). Schlüssel müssen unveränderlich sein und können Datentypen wie Strings, Zahlen oder Booleans usw. sein.

- Item: Dies ist der Wert, der dem Schlüssel zugeordnet wird (z.B. "Item1"). Ein Wert kann jeder beliebige Datentyp sein, wie z.B. ein String, eine Zahl, eine Liste, ein weiteres Dictionary, etc. 

- *Mehrer Werte können in einer Liste [**strings, booleans, Floats, Integers**] eingesetzt werden das sichert die einfache Manipulation, filterung der Daten.*

In [ ]:
Lagerbestand = {"Früchte":["Erdbeere", "Apfel", "Birne"],
                "Menge": [40, 30, 0],
                "Verfügbarkeit": [True, True, False]}
                

Lagerbestand

#### Umganz zur Daten Manipulation mit Dictionaries

- Eine Spalte (Key) hinzufügen mit Werte (Item)

Im darunter liegenden Code wird zur der bestehenden Variable `Lagerbestand`
die Spalte `Verfügbarkeit` mit einer `liste` mit Werten hinzugefügt.
Welche Werte du mitgibst ist die überlassen ob es Booleans, Integers, Floats, Strings sind.


In [ ]:
 # <- Liste
Lagerbestand['Herkunftsland'] = ["Deutschland", "Frankreich", "Dänemark"]

print("Dein Aktualisiertes Dictionary:")
Lagerbestand

#### **auf Werte zugereifen**
Wenn du ein Dictionary hast mit vielen Keys uns Items kannst du auch nur deine Wunsch Key extrahieren und betrachten dies erfolgt wie folgt:

In [ ]:
Lagerbestand['Früchte']  # Nur die Früchte werden dir angezeigt


#### **Keys entfernen**

- `del` entfernt den key und die items deines Dictionary

In [23]:
del Lagerbestand["Herkunftsland"] 

#### **Auf keys zugreifen**

- Hier wird nach den Keys gefiltert die Ausgabe sind nur die Keys ohne die Items

In [ ]:
print(Lagerbestand.keys())

#### **auf Iteams zugreifen**
Zu dem jeweiligen Key werden nur die Items herausgegeben

In [ ]:
print(Lagerbestand.get("Früchte"))


#### Alle items wiedergeben
`.values()`gibt dir alle Items in jeden Keys wieder


In [ ]:
print(Lagerbestand.values())

#### Neue werte hinzufügen
durch den `Key`kannst du 

In [27]:
Lagerbestand["Früchte"].append("Kirsche")
Lagerbestand["Menge"].append(30)
Lagerbestand["Verfügbarkeit"].append(True)


In [ ]:
Lagerbestand

## 1.6 Schleifen 
---

Loops (Schleifen) in Python ermöglichen es, bestimmte Codeblöcke wiederholt auszuführen. Die beiden gängigsten Arten von Schleifen sind die while-Schleife und die for-Schleife. Mit der **while**-Schleife wird der Codeblock ausgeführt, solange eine angegebene Bedingung erfüllt ist. Mit der **for**-Schleife wird der Codeblock für jedes Element in einer Sequenz, wie z. B. einer Liste, wiederholt ausgeführt.

### 1.6.1 For-Schleife
`for`-Schleifen werden verwendet, um eine Sequenz schrittweise zu durchlaufen, wie zum Beispiel eine Liste, String oder ein Array.


- Eine `for`-Schleife kann verwendet werden, um die Elemente einer Liste einzeln zu durchlaufen und zu verarbeiten.

- Betrachte das folgende Beispiel, bei dem die Liste früchte durchlaufen wird. Die Liste enthält die Elemente: 

***Früchte = ['Apfel', 'Kiwi', 'Banane', 'Kirsche', 'Orange', 'Ananas'].***

- Die `for`-Schleife hilft uns dabei, jedes Element der Liste nacheinander zu extrahieren und zu verwenden.

**Grundlegende Syntax einer `for`-Schleife:**


Übliche Syntax für ein Schleifen (loops):

In [ ]:
for frucht in Früchte:
    print(frucht)

**Achtung:**
- Der Datentyp `Früchte` ist eine Liste (`list`).

- Die `for`-Schleife durchläuft alle Elemente der Liste und gibt sie als Strings aus.
- Sollte es eine `float` oder `int` werden dann dementsprechen `int` oder `float` als Datentypen angepasst


### 1.6.2 While-Schleife
---
Wie in Punk 1.6.1 bereits beschrieben wird die `while-schleife` solange durch geführt bis die definierte Bedigung erfüllt wird.

1. In diesem Code-Snippet wird eine `while-Schleife` verwendet, um die Zahlen von 1 bis 5 auszugeben. Zuerst wird die Variable i auf den Wert 1 gesetzt. 
2. Die Schleife prüft dann, ob `i` kleiner als 6 ist. Solange das der Fall ist, wird der Wert von `i` ausgegeben und anschließend um 1 erhöht. 
3. Dieser Vorgang wiederholt sich, bis `i` den Wert 6 erreicht, wodurch die Schleife endet. Das Ergebnis dieses Codes ist eine aufeinanderfolgende Ausgabe der Zahlen 1 bis 5.

In [ ]:
i = 1
while i < 6:
  print(i)
  i += 1

# 1.7 if, elif, else Bedigungen

### 1.7.1 If-Bedigungen
Bei einer If Bedigung wird geprüft ob die gegbene Bedigung erfüllt wird. Wenn diese erfüllt wird
wird nur der Code unter der If-Bedigung ausgeführt.

In [ ]:
Preis = 10 # <- hier kannst du ein belibigen Preis eingeben

if Preis == 10:
    print("Der Preis ist 10 €")

### 1.7.2 if else-Bedigungen

Im nächsten Szenario wird gepruft ob die If-Bedigung wahr ist, wenn die bedigung erfüllt wird führt Sie nur die If-Bedigung.
Ist jedoch die If-Bedigung **nicht** erfüllt dann wird **immer** die else-Bedigung erfüllt.


**ACHTUNG!**

Die else bedigung ist mit vorsicht zu nutzen. Sie wird IMMER ausgeführt sollte die If-Bedigung nicht erfüllt werden.
Wenn in deinen Code nur eine IF und ELSE bedigung genutzt wird, wird die ELSE bedigung immer ausgeführt dies kann in gewissen Situationen zu einem unerfwünschten ergebnis führen.


- **Daher wird nun zusätzlich die elif bedigung betrachtet.**


In [ ]:
Preis = 22 # <- hier kannst du ein belibigen Preis eingeben

if Preis == 10:
    print("Der Preis ist 10 €")
else:
    print("Der Preis beträgt alles andere, aber nicht 10 €")

### 1.7.3 if elif else Bedingungen

Die elif-Bedingung ist eine weitere Bedingung, die wie die if-Bedingung nur dann ausgeführt wird, wenn die Bedingung erfüllt ist.
Das bedeutet, dass zuerst geprüft wird, ob die if-Bedingung erfüllt ist. Wenn nicht, wird geprüft, ob die elif-Bedingung erfüllt ist. Wenn ja, wird dieser Code ausgeführt.

- Elif-Bedingungen können so oft erstellt werden, wie notwendig. Beachte aber, dass der Code verständlich und nicht unnötig verkompliziert wird.



In [ ]:
Preis = 30 # <- hier kannst du ein belibigen Preis eingeben

if Preis == 10:
    print("Der Preis ist 10 €")
elif Preis == 20:
    print("Der Preis ist 20 €")
elif Preis > 10:
    print("Der Preis ist mehr als 10 €")
elif Preis < 10:
    print("Der Preis ist weniger als 10 €")
else:
    print("Der Preis ist mehr als 20 €")

# 1.8 Funktionen
---
Eine Funktion kann man als ein wiederverwendbares Paket aus Anweisungen sehen. 
Stellen wir uns vor, du möchtest in einem Programm eine bestimmte Aufgabe immer wieder erledigen – zum Beispiel, eine Zahl zu addieren. Anstatt den gleichen Code immer wieder zu schreiben, kannst du diesen in eine Funktion packen und ihn so oft verwenden, wie du möchtest.

Wie umfangreich deine Funktion wird ist dir überlassen.

**Kleiner Tipp**: 
Beim schreiben einer Funktion füge ein Beschreibung hinzu sowie was die Funktion erwartet Input und Output.
**KURZ UND PRÄGNAT KEINE ROMANE!**

In [ ]:
def addiere(zahl1: int, zahl2: int):

    """
    Addiert zwei Zahlen miteinander

    Input: Zwei Integers
    Output : Die addition der beiden Zahlen
    """ 
    summe = zahl1 + zahl2
    return summe

addiere(4,1) #<- trage hier zahlen zur addition 

In [ ]:
addiere(32,23)

# 2. Libraries
----

Was ist eine Python-Bibliothek (Libary)?

Eine Bibliothek ist eine Sammlung von Code, die alltägliche Aufgaben effizienter macht. 

- Mit **Pandas** kann man Dataframes Manipulieren, Zeilen hinzufügen, Daten bereiningung und vieles mehr.
- Mit **Plotly** können mit wenigen Zeilen Code ganze Graphen von Linien, Balken bis hinzu Kerzendiagramme erstellt werden.
- Mit **Scikit-learn** wird genutzt für das maschinellen Lernen. Es bietet verschiedene Klassifikations-, Regressions- und Clustering-Algorithmen-
- Mit **Streamlit** kannst du Problemlos Webapplikation erstellen werden

## 2.1. Pandas
---
Pandas ist eine leistungsstarke Open-Source-Bibliothek in Python, die Werkzeuge für die Arbeit mit strukturierten Daten bietet. Sie ist besonders nützlich für Datenanalyse und -manipulation, da sie einfache, aber effiziente Datenstrukturen wie DataFrames bereitstellt.


Installation
Bevor man Pandas verwenden kann, muss die Library installiert werden:

`pip install pandas`

In [ ]:
!pip install pandas

Verwendung
Nachdem Pandas installiert ist, kann es in Python wie folgt importiert werden:

In [ ]:
import pandas as pd

Um Daten zu importieren nutzen wir `pd.read_csv()`
mit dem befehl help(pd.read_csv) bekommen wir Hilfestellung wie wir die Funktion nutzen können.

In [ ]:
help(pd.read_csv)

Wenn die Anleitung aufmerksam gelesen wird ist der erster Parameter:

`filepath_or_buffer = str, path object or file-like object`

"Any valid string path is acceptable."
Das bedeutet gib den Parameter wo deine Datei liegt indem Fall eine comma seprated Value (.csv) als String(str),

- ***filepath_or_buffer: "Desktop/Daten/dein_Datei_namem.csv"***


`Seprator - sep: str, default ‘,’`
Weiterhin um die Datei richtig einzulesen muss man ein Seprator mit eingeben in unseren Fall ist es eine comma seprated Value (.csv) als Dateiformat also macht es Sinn ein Komme zusetzten als String.

- ***sep = ","***

Zusammengefasst wäre der Code wie folgt:

In [ ]:
pd.read_csv(filepath_or_buffer = "Desktop/Daten/deine_Datei_Namen_einfügen.csv",
            sep=",")

In [52]:
df = pd.read_csv(filepath_or_buffer = "Daten/house_price_regression_dataset.csv",
            sep=",") 

- Beachte hier wird die CSV Tabelle in die Variable df eingespeichert. Somit müssen wir den ganze Code nicht immer durch führen

df.head():

Ist eine weitere Pandas Funktion um die Daten zubetrachten. Dabei werden die ersten 5 Zeilen angezeigt

Sollte hier in die df.head() Funktion ein integer mitgeben werden wie 20 werden dementsprechend 20 Zeilen betrachtet.

In [54]:
df.head()           # <- Die ersten 5 Zeilen
# df.head(20)       # <- Die ersten 20 Zeilen

Square_Footage  Num_Bedrooms  Num_Bathrooms  Year_Built  Lot_Size  \
0            1360             2              1        1981  0.599637   
1            4272             3              3        2016  4.753014   
2            3592             1              2        2016  3.634823   
3             966             1              2        1977  2.730667   
4            4926             2              1        1993  4.699073   

   Garage_Size  Neighborhood_Quality   House_Price  
0            0                     5  2.623829e+05  
1            1                     6  9.852609e+05  
2            0                     9  7.779774e+05  
3            1                     8  2.296989e+05  
4            0                     8  1.041741e+06

Um ein besseres Verständis für deine Daten zubekommen hilft die Funktion:
`df.describe()`

In [58]:
df.describe()

Square_Footage  Num_Bedrooms  Num_Bathrooms   Year_Built     Lot_Size  \
count     1000.000000   1000.000000    1000.000000  1000.000000  1000.000000   
mean      2815.422000      2.990000       1.973000  1986.550000     2.778087   
std       1255.514921      1.427564       0.820332    20.632916     1.297903   
min        503.000000      1.000000       1.000000  1950.000000     0.506058   
25%       1749.500000      2.000000       1.000000  1969.000000     1.665946   
50%       2862.500000      3.000000       2.000000  1986.000000     2.809740   
75%       3849.500000      4.000000       3.000000  2004.250000     3.923317   
max       4999.000000      5.000000       3.000000  2022.000000     4.989303   

       Garage_Size  Neighborhood_Quality   House_Price  
count  1000.000000           1000.000000  1.000000e+03  
mean      1.022000              5.615000  6.188610e+05  
std       0.814973              2.887059  2.535681e+05  
min       0.000000              1.000000  1.116269e+05  
25%       0.000000              3.000000  4.016482e+05  
50%       1.000000              6.000000  6.282673e+05  
75%       2.000000              8.000000  8.271413e+05  
max       2.000000             10.000000  1.108237e+06

Der Datensatz enthält Informationen über **1000 Immobilien**.

- Die **durchschnittliche Wohnfläche** beträgt etwa **2815 Quadratfuß**, wobei die kleinste Immobilie **503 Quadratfuß** groß ist und die größte **4999 Quadratfuß** misst.
- Im Durchschnitt haben die Immobilien **3 Schlafzimmer** und **2 Badezimmer**. Die Anzahl der Schlafzimmer variiert zwischen **1 und 5**, und die Anzahl der Badezimmer liegt zwischen **1 und 3**.
- Das **durchschnittliche Baujahr** der Immobilien liegt bei **1986**, wobei die ältesten Immobilien im Jahr **1950** gebaut wurden und die neuesten im Jahr **2022**.
- Die **durchschnittliche Grundstücksgröße** beträgt **2,78 Acres**, mit einem Minimum von **0,51 Acres** und einem Maximum von **4,99 Acres**.
- Im Durchschnitt haben die Immobilien eine **Garagengröße von 1 Auto**, wobei einige Immobilien **keine Garage** haben und andere Platz für bis zu **2 Autos** bieten.
- Die **Nachbarschaftsqualität** liegt im Durchschnitt bei **5,6** auf einer Skala von **1 bis 10**, wobei die niedrigste Qualität **1** und die höchste **10** beträgt.
- Der **durchschnittliche Hauspreis** liegt bei etwa **618.861 USD**, mit einem Minimum von **111.627 USD** und einem Maximum von **1.108.237 USD**.
- Die mittleren **50% der Hauspreise** bewegen sich zwischen **401.648 USD** und **827.141 USD**, was auf eine große Preisspanne hinweist.


In [66]:
#!pip install plotly
!pip install nbformat==4.2.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.3/153.3 kB 2.1 MB/s eta 0:00:00a 0:00:01
  Attempting uninstall: nbformat
    Found existing installation: nbformat 5.10.4
    Uninstalling nbformat-5.10.4:
      Successfully uninstalled nbformat-5.10.4

[notice] A new release of pip available: 22.3.1 -> 24.2
[notice] To update, run: pip install --upgrade pip


In [61]:
import plotly.express as px
df.columns

Index(['Square_Footage', 'Num_Bedrooms', 'Num_Bathrooms', 'Year_Built',
       'Lot_Size', 'Garage_Size', 'Neighborhood_Quality', 'House_Price'],
      dtype='object')

In [67]:
fig = px.line(df, x="Year_Built", y="Square_Footage")
fig.show()

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed